local-tests/chainrunner/node.py

```diff
@@ -0,0 +1,92 @@
+import json
+import os.path as op
+import re
+import requests
+import jsonrpcclient as rpc
+import subprocess
+
+from .utils import flags_from_dict
+
+class node:
+    """A class representing a single node of a running blockchain."""
+    def __init__(self, binary, chainspec, path, logdir=None):
+        self.chainspec = chainspec
+        self.binary = binary
+        self.path = path
+        self.logdir = logdir or path
+        self.logfile = None
+        self.flags = {}
+        self.running = False
+
+
+    def _stdargs(self):
+        return ['--base-path', self.path, '--chain', self.chainspec]
+
+
+    def start(self, name):
+        """Start the node. `name` is used to name of the logfile and for --name flag."""
+        cmd = [self.binary, '--name', name] + self._stdargs() + flags_from_dict(self.flags)
+        self.logfile = op.join(self.logdir, name+'.log')
+        self.process = subprocess.Popen(cmd, stderr=open(self.logfile, 'w'), stdout=subprocess.DEVNULL)
+        self.running = True
+
+
+    def stop(self):
+        """Stop the node by sending SIGKILL."""
+        self.process.kill()
+        self.running = False
+
+
+    def purge(self):
+        """Purge chain (delete the database of the node)."""
+        cmd = [self.binary, 'purge-chain', '-y'] + self._stdargs()
+        subprocess.run(cmd, stdout=subprocess.DEVNULL).check_returncode()
+
+
+    def greplog(self, regexp):
+        """Find in the logs all occurences of the given regexp. Returns a list of matches."""
+        if self.logfile:
```

Comment:
remove nesting by:
`if not self.logfile: return []`

Reply:
Done